File: cli/insertBuild.rb
require 'optparse'
require 'fileutils'
require 'digest/sha2'
require 'rugged'
require 'mongo'

$standard_channels = %w[DEFAULT EXPERIMENTAL]
$standard_types = ["application"]

class Download

  attr_reader :type, :hash, :path

  def initialize(type, hash, path)
    @type = type
    @hash = hash
    @path = path
  end

end

def parse_arguments
  options = {}
  options[:downloads] = []
  OptionParser.new do |option|
    option.on("--projectName Name") { |name|
      options[:project_name] = name
    }
    option.on("--projectFriendlyName NAME") { |name|
      options[:project_friendly_name] = name
    }
    option.on("--versionGroup GROUP") { |group|
      options[:group] = group
    }
    option.on("--version VERSION") { |version|
      options[:version] = version
    }
    option.on("--build BUILD, AUTO") { |build|
      if build == "AUTO"
        options[:build_auto] = true
        options[:build] = nil
      else
        options[:build] = build
      end
    }
    option.on("--channel DEFAULT, EXPERIMENTAL") { |channel|
      unless $standard_channels.include?(channel)
        abort("[ERROR] Unknown channel: #{channel} please specify a valid channel#{$standard_channels}")
      end
      options[:channel] = channel
    }
    option.on("--repo PATH") { |path|
      options[:repo] = path
    }
    option.on("--storage PATH") { |path|
      options[:storage] = path
    }
    option.on("--download TYPE:PATH:HASH, TYPE:PATH") { |download|
      data = download.split(':')
      unless data.length >= 2
        abort("[ERROR] Invalid data length for download '#{data[0]}'")
      end

      type = data[0]
      unless $standard_types.include?(type)
        puts "[WARNING] You are not using[#{type}] standard types please consider using one of those: #{$standard_types}"
      end

      path = data[1]
      if File.exist?(path)
        if data.length > 2
          hash = data[2]
        else
          puts "[INFO/hash] Creating SHA256 hash for file '#{path}'"
          hash = Digest::SHA2.hexdigest(File.read(path))
          puts "[INFO/hash] Hash: #{hash}"
        end
      else
        abort("[ERROR] Could not find file '#{path}'")
      end

      options[:downloads].append(Download.new(type, hash, path))
    }
  end.parse!
  options
end

def check_arguments(options)
  unless options.has_key?(:project_name)
    abort("[ERROR] Please specify a project ID using --projectName. Example: --projectName raper")
  end
  unless options.has_key?(:project_friendly_name)
    abort("[ERROR] Please specify a project name using --projectFriendlyName. Example: --projectFriendlyName Raper")
  end
  unless options.has_key?(:group)
    abort("[ERROR] Please specify a version group using --versionGroup. Example: --versionGroup 1.20")
  end
  unless options.has_key?(:version)
    abort("[ERROR] Please specify a version using --version. Example: --version 1.20.1")
  end
  unless options.has_key?(:build)
    abort("[ERROR] Please specify a build number using --build. Example: --build 1")
  end
  unless options.has_key?(:channel)
    options[:channel] = "DEFAULT"
  end
  unless options.has_key?(:repo)
    abort("[ERROR] Please specify a repository path using --repo. Example: --repo raper/")
  end
  unless options.has_key?(:storage)
    abort("[ERROR] Please specify a storage path using --storage. Example: --storage /home/api/storage/")
  end
  if options[:downloads].length == 0
    abort("[ERROR] Please specify one or more downloads using --download. Example: --download application:paper-1.20.jar")
  end
end

def connect_to_mongodb
  puts "[INFO/database] Trying to establish connection to MongoDB Server..."
  url = "mongodb://127.0.0.1/bibliothek"
  if File.exist?("mongodb.url")
    url = File.read("mongodb.url")
  else
    puts "[INFO/database] No mongodb.url found using default #{url}"
  end
  Mongo::Client.new(url)
end

# Main
options = parse_arguments
build_number = options[:build].to_i
check_arguments options

client = connect_to_mongodb

puts "[INFO/database] Fetching and creating information..."
project = client[:projects].find_one_and_update(
  {
    "name": options[:project_name]
  },
  {
    "$setOnInsert": {
      "name": options[:project_name],
      "friendlyName": options[:project_friendly_name],
    }
  },
  :upsert => true,
  :return_document => :after
)
group = client[:groups].find_one_and_update(
  {
    "project": project["_id"],
    "name": options[:group]
  },
  {
    "$setOnInsert": {
      "project": project["_id"],
      "name": options[:group],
      "timestamp": Time.now
    }
  },
  :upsert => true,
  :return_document => :after
)
version = client[:versions].find_one_and_update(
  {
    "project": project["_id"],
    "name": options[:version]
  },
  {
    "$setOnInsert": {
      "project": project["_id"],
      "group": group["_id"],
      "name": options[:version],
      "timestamp": Time.now
    }
  },
  :upsert => true,
  :return_document => :after
)

puts "[INFO/build] Generating build number..."
previous_build = client[:builds].find(
  {
    "project": project["_id"],
    "version": version["_id"]
  },
  {
    "sort": {
      "_id": -1
    }
  }
).limit(1).first

if options.has_key?(:build_auto)
  if previous_build.nil?
    build_number = 1
    puts "[INFO/build] Next build number is 1"
  else
    build_number = previous_build["number"] + 1
    puts "[INFO/build] Next build number is #{build_number}"
  end
end

puts "[INFO/git] Loading git repo"
repo = Rugged::Repository.new(options[:repo])
commits = []

if !previous_build.nil? && previous_build[:changes].length > 0
  last_commit = repo.lookup(previous_build[:changes][0][:commit])
else
  last_commit = repo.head.target.parents.first
end

walker = Rugged::Walker.new(repo)
walker.sorting(Rugged::SORT_TOPO)
walker.push(last_commit)
walker.each do |commit|
  puts "[INFO/git] Commit: #{commit.message.strip!} # #{commit.oid}"

  commits << {
    "commit": commit.oid,
    "summary": commit.message,
    "message": commit.message
  }

  break if commit.oid == last_commit.oid
end

downloads = {}
options[:downloads].each do |download|
  downloads[download.type] = {
    "name": options[:project_name] + "-" + options[:version] + "-" + build_number.to_s + ".jar",
    "sha256": download.hash
  }
end

downloads_path = File.join(options[:storage], options[:project_name], options[:version], build_number.to_s)
puts "[INFO/files] Creating hosting directory: #{downloads_path}"
FileUtils.mkdir_p(downloads_path)

options[:downloads].each do |download|
  download_path = File.join(downloads_path, options[:project_name] + "-" + options[:version] + "-" + build_number.to_s + ".jar")
  puts "[INFO/files] Copying #{download.path} to #{download_path}"
  FileUtils.copy_file(download.path, download_path)
end

puts "[INFO/database] Adding build to database..."
client[:builds].insert_one(
  {
    "project": project["_id"],
    "version": version["_id"],
    "number": build_number,
    "timestamp": Time.now,
    "changes": commits,
    "downloads": downloads,
    "displayMode": "HIDE",
    "channel": options[:channel]
  }
)
puts "[FINISHED] Inserted build #{build_number.to_s} (Channel: #{options[:channel]}) for project #{options[:project_name]} (#{project["_id"]}) version #{options[:version]} (#{version["_id"]})"

File: cli/insertBuild.js
const FileSystem = require("fs");
const Path = require("path");
const YArgs = require("yargs");
const GitLog = require("gitlog").default;
const {MongoClient} = require("mongodb");

const standardChannels = ["DEFAULT", "EXPERIMENTAL"];
const standardTypes = ["application"];

class Download {

    constructor(type, path, hash, name = null) {
        this.type = type;
        this.path = path;
        this.hash = hash;
        this.name = name;
    }

    valid() {
        return FileSystem.existsSync(this.path);
    }

}

function parseArguments() {
    return YArgs
        .option("projectName", buildOptions("p", "Example: raper"))
        .option("projectFriendlyName", buildOptions("n", "Example: Raper"))
        .option("versionGroup", buildOptions("g", "Example: 1.20"))
        .option("versionName", buildOptions("v", "Example: 1.20.1"))
        .option("buildNumber", buildOptions("b", "Example: -1 for AUTO or NUMBER", true, "number"))
        .option("repositoryPath", buildOptions("r", "Example: raper/"))
        .option("storagePath", buildOptions("s", "Example: storage/"))
        .option("download", buildOptions("d", "Example: application:buildJar.jar:hash:name"))
        .option("buildChannel", buildOptions("c", "Example: default or experimental", false))
        .default("buildChannel", "default")
        .help()
        .version(false)
        .argv;
}

function checkOptions(options) {
    options.buildChannel = options.buildChannel.toUpperCase(); // Build to upper case
    options.buildNumber = parseInt(options.buildNumber);

    if(!standardChannels.includes(options.buildChannel)) {
        console.error(`[ERROR] Unknown channel: "${options.buildChannel}" please specify a valid channel: ${standardChannels.join(", ")}`);
        process.exit(1);
    }

    if(typeof options.download === "string") {
        options.download = [options.download];
    }

    let primary = false;
    let downloads = [];
    options.download.forEach((download) => {
        const data = download.split(":");
        if(data.length >= 4) {
            download = new Download(data[0], data[1], data[2], data[3]);
        } else if(data.length === 3) {
            if(primary) {
                console.error("[ERROR] To many primary files, please check that!");
                process.exit(1);
            }

            download = new Download(data[0], data[1], data[2], null);
            primary = true;
        }
        if(!download.valid()) {
            console.error(`[ERROR] Invalid download found: ${download.type}`);
            download = null;
        }

        if(download) {
            console.log(`[INFO/download] Validated and added download ${download.type}`);
            downloads.push(download);
        }
    });
    if(downloads.length <= 0) {
        console.error("[ERROR] No downloads left. Quitting...");
        process.exit(1);
    }
    options.download = downloads;

}

function buildOptions(alias, description, required = true, type = "string") {
    return {
        alias: alias,
        describe: description,
        required: required,
        type: type
    }
}

async function writeOrGetProject(database, name, friendlyName) {
    return database.collection("projects").findOneAndUpdate(
        { name: name },
        {
          $setOnInsert: {
            name: name,
            friendlyName: friendlyName
          }
        },
        { upsert: true, returnOriginal: false }
    );
}

async function writeOrGetGroup(database, project, name) {
    return database.collection("groups").findOneAndUpdate(
        { project: project.value._id, name: name },
        {
          $setOnInsert: {
            project: project.value._id,
            name: name,
            timestamp: new Date()
          }
        },
        { upsert: true, returnOriginal: false }
    );
}

async function writeOrGetVersion(database, project, group, name) {
    return database.collection("versions").findOneAndUpdate(
        { project: project.value._id, name: name },
        {
          $setOnInsert: {
            project: project.value._id,
            group: group.value._id,
            name: name,
            timestamp: new Date()
          }
        },
        { upsert: true, returnOriginal: false }
    );
}

async function writeBuild(database, project, version, buildNumber, channel, changes, downloads) {
    return database.collection('builds').insertOne({
        project: project.value._id,
        version: version.value._id,
        number: buildNumber,
        timestamp: new Date(),
        changes: changes,
        downloads,
        displayMode: 'HIDE',
        channel: channel
    });
}

async function findLatestBuild(database, project, version) {
    return database.collection("builds").findOne(
        { project: project.value._id, version: version.value._id },
        { sort: { _id: -1 } }
    );
}

async function connectToMongoDB() {
    console.log("[INFO/database] Trying to establish connection to MongoDB Server...");

    let url = "mongodb://127.0.0.1/bibliothek";
    if(process.env.MONGODB_URL) {
        url = process.env.MONGODB_URL;
    } else if(FileSystem.existsSync("mongodb.url")) {
        url = FileSystem.readFileSync("mongodb.url", "utf-8");
    } else {
        console.log(`[INFO/database] No environment variable or mongodb.url file found. Using default: ${url}`);
    }

    const client = new MongoClient(url);

    try {
        await client.connect();
        return client;
    } catch(error) {
        console.error(`[ERROR/database] Failed to connect to MongoDB: ${error}`);
        process.exit(1);
    }
}

function buildStoragePath(storagePath, projectName, versionName, buildNumber) {
    return Path.join(storagePath, projectName, versionName, buildNumber.toString());
}

function generateName(download, projectName, versionName, buildNumber) {
    return download.name ? download.name : projectName + "-" + versionName + "-" + buildNumber + ".jar";
}

function copyDownload(download, storageFolder, projectName, versionName, buildNumber) {
    FileSystem.copyFileSync(download.path, Path.join(storageFolder, generateName(download, projectName, versionName, buildNumber)));
}

async function main() {
    const options = parseArguments();
    checkOptions(options);

    buildNumber = options.buildNumber;
    options.buildNumber = null; // Prevent usage

    const client = await connectToMongoDB();
    const database = client.db();

    console.log("[INFO/database] Fetching and creating information...");
    const project = await writeOrGetProject(database, options.projectName, options.projectFriendlyName);
    const group = await writeOrGetGroup(database, project, options.versionGroup);
    const version = await writeOrGetVersion(database, project, group, options.versionName);

    const previousBuild = await findLatestBuild(database, project, version);

    if(buildNumber === -1) {
        if(previousBuild) {
            buildNumber = previousBuild.number + 1;
        } else {
            buildNumber = 1;
        }
        console.log(`[INFO/build] Next build number is ${buildNumber}`);
    }

    // BuildNumber should be set for now. Ready to copy files

    // Copy files
    console.log("[INFO/copy] Copying files...");
    storagePath = buildStoragePath(options.storagePath, options.projectName, options.versionName, buildNumber);
    if(!FileSystem.existsSync(storagePath)) {
        FileSystem.mkdirSync(storagePath, { recursive: true });
    }
    options.download.forEach((download) => {
        copyDownload(download, storagePath, options.projectName, options.versionName, buildNumber);
    });
    console.log("[INFO/copy] Finished!");

    const lastBuildCommit = previousBuild && previousBuild.changes.length ? previousBuild.changes.slice(0, 1)[0].commit : "HEAD^1";

    // Get changes since last build
    let changes = [];
    GitLog(
        {
            repo: options.repositoryPath,
            fields: [ "hash", "subject", "rawBody" ],
            branch: lastBuildCommit + "...HEAD"
        }
    ).forEach((commit) => {
        changes.push(
            {
                commit: commit.hash,
                summary: commit.subject,
                message: commit.rawBody
            }
        );
        console.log(`[INFO/git] Commit: ${commit.subject} # ${commit.hash}`);
    });

    // Create downloads
    const downloads = {};
    options.download.forEach((download) => {
        downloads[download.type.replace(".", ":")] = {
            name: generateName(download, options.projectName, options.versionName, buildNumber),
            sha256: download.hash
        }
    });

    console.log("[INFO/database] Adding build to database...");
    await writeBuild(database, project, version, buildNumber, options.buildChannel, changes, downloads);

    await client.close();
    console.log(`[FINISHED] Inserted build ${buildNumber} (Channel: ${options.buildChannel}) for project ${options.projectName} (${project.value._id}) version ${options.versionName} (${version.value._id})`);
}

main().catch((error) => {
    console.error(`[ERROR] An unhandled exception occurred: ${error}`);
    process.exit(1);
});